src/lib.rs

```diff
@@ -198,71 +198,142 @@ fn is_wasm_module(path: impl AsRef<Path>) -> Result<bool> {
     Ok(bytes[4..] == [0x01, 0x00, 0x00, 0x00])
 }
 
-fn create_component(config: &Config, target_path: &Path) -> Result<()> {
-    let dep_path = target_path
+fn create_component(config: &Config, output: &UnitOutput, binary: bool) -> Result<()> {
+    let deps_dir = output
+        .path
         .parent()
-        .unwrap()
-        .join("deps")
-        .join(target_path.file_name().unwrap());
+        .expect("expected the output to have a parent directory")
+        .join("deps");
+
+    // Determine what the file would be in the deps directory
+    // We need to do this because cargo will *always* link/copy the dep to
+    // the output directory, even when it's up-to-date.
```

Comment:
I did some quick tests with strace on `cargo-component build` and am not seeing this. I see the `linkat` on the first cargo-component-build run and not subsequent runs when the output dir is up to date. Can you say more about where you're seeing this?

Reply:
So there are some differences based on platform for what cargo does here.

On Linux, it's probably doing a hardlink and the inodes are the same, so cargo skips making a syscall (there's a test in there if the deps and target paths are the "same" file that bails out successfully).

On macOS, where I am seeing the behavior, it's always doing a copy because of a workaround for an AFS bug. So the inodes differ and cargo always copies the deps over the target even when they're the "same". Because of this, if we don't componentize the deps file, cargo will overwrite the previous component with the module output by the Rust toolchain, which will trigger `cargo component` to componentize it again (and printing a message that it's doing so).

Reply:
I'll at least update the comment to mention that this behavior appears to be macOS-specific.

Reply:
Ok. This seems ok for now, though it feels the kind of thing someone may end up revisiting in the future. Users will hit errors because when they end up with multiple versions in their deps directory, and it's possible we may hit other cargo implementation details. So the more we can record about what we know the situation in the comments, the better :-).

Reply:
Actually, you're right, this isn't a good way of doing this for binaries as cargo will indeed leave the stale ones there after `Cargo.toml` changes, so we shouldn't do it this way.

Let me see if I can figure out exactly what file hash we should be using so we don't search for it.

Reply:
Alright, there's no good way to do it from cargo's API: the fingerprint calculation is done internal to the compile jobs and doesn't propagate to the compilation result.

I think I'm going to modify this to simply write to the target path, foregoing any deps writing. Doing so also makes it less reliant on cargo internals to make work.

The unfortunate side effect is that, on macOS, users will always see a "creating component" message even if the build was up-to-date.

Perhaps we can work around that by caching mtimes of the files pre-componentization to compare against and forego the message if the mtime matches what was cached previously; we'll still need to componentize since cargo is copying from deps over the previous componentized target, though.

That's a fix for another day.